src/go/rpk/pkg/cli/cmd/root_test.go

```diff
@@ -14,3 +18,123 @@ func TestLogOutput(t *testing.T) {
 	// STDOUT.
 	require.Exactly(t, os.Stdout, logrus.StandardLogger().Out)
 }
+
+func TestTryExecPlugin(t *testing.T) {
+	fs := func() testPluginHandler {
+		return testPluginHandler{
+			"rpk-foo_bar-baz":     nil,
+			"rpk-foo_bar":         nil,
+			"rpk-fizz-buzz-bizzy": nil,
+		}
+	}
+	hit := func(file, args string) testPluginHandler {
+		h := fs()
+		h["rpk-"+file] = map[string]int{args: 1}
+		return h
+	}
+
+	for _, test := range []struct {
+		name      string
+		args      []string
+		exp       testPluginHandler
+		unhandled bool
+	}{
+		{
+			name: "prefer longest command match",
+			args: []string{"foo-bar", "baz"},
+			exp:  hit("foo_bar-baz", ""),
+		},
+
+		{
+			name: "match shorter command if not all pieces specified",
+			args: []string{"foo-bar", "--baz"},
+			exp:  hit("foo_bar", "--baz"),
+		},
+
+		{
+			name: "match shorter command, flag with argument",
+			args: []string{"foo-bar", "--baz", "buzz"},
+			exp:  hit("foo_bar", "--baz\x00buzz"),
+		},
+
+		{
+			name: "match shorter command, no flags",
+			args: []string{"foo-bar"},
+			exp:  hit("foo_bar", ""),
+		},
+
+		{
+			name: "pieces joined with dash",
+			args: []string{"fizz", "buzz", "bizzy"},
+			exp:  hit("fizz-buzz-bizzy", ""),
+		},
+
+		{
+			name:      "no plugin found",
+			args:      []string{"unknown", "unknown", "unknown"},
+			exp:       fs(),
+			unhandled: true,
+		},
+
+		{
+			name:      "exec error returns error",
+			args:      []string{testPluginFileExecError},
+			exp:       fs(),
+			unhandled: true,
+		},
+
+		{
+			name:      "no args is unhandled",
+			args:      []string{},
+			exp:       fs(),
+			unhandled: true,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			fs := fs()
+			foundPath, err := tryExecPlugin(fs, test.args)
+			if err != nil {
+				if len(test.args) > 0 && test.args[1] != "ERROR" {
+					t.Errorf("unexpected tryExecPlugin error for arg0 %s: %v", test.args[0], err)
+				}
+				return
+			}
+			if !test.unhandled && len(foundPath) == 0 {
+				t.Error("foundPath is unexpectedly empty")
+			}
+			if diff := cmp.Diff(fs, test.exp); diff != "" {
+				t.Errorf("%s: %s", test.name, diff)
```

Comment:
@twmb we use `require` for assertions: https://github.com/stretchr/testify

It has good default messages, and comes in handy for stuff like this^

Reply:
Assertions aren't really idiomatic Go, and I've avoided that library & assertions in general since 2012.
Ian Lance Taylor describes TDD [here](https://github.com/golang/go/wiki/TableDrivenTests) and importantly references the Go FAQ about why Go does not have assertions ([here](https://golang.org/doc/faq#assertions)). Conversely, `go-cmp` is so standard that the author is a core Go contributor and has a proposal open to include it in the stdlib ([here][https://github.com/golang/go/issues/45200]).

Reply:
If I understand it correctly, the argument is against language support for assertion statements, like Python's `assert`. I agree with that statement's sentiment, by the way. If there's anything I like about Go is that it makes handling errors explicitly easy.

But having functions to compare values, their types & structure has proved to be convenient. So far we haven't found any terrible downsides to using `require`.